### DISCUSSION TITLE
Distribution of Customized Apple Wallet Passes After Generation Using PassKit-Generator

### DISCUSSION THREAD

**Discussion by akash-86** [alexandercerutti/passkit-generator#183] [Q&A]
I'm currently working on implementing an Apple Wallet system for my web application, and I've successfully generated personalized event tickets using PassKit-Generator. Now, I'm seeking guidance on the subsequent steps for distributing these passes to users when they click on an "Add to Apple Wallet" button. Specifically, I'd appreciate insights on how to seamlessly enable users to add these passes to their Apple Wallets after the passes have been generated. Any advice, code snippets, or recommended libraries related to pass distribution and integration with Apple Wallet would be highly valuable. Your expertise is greatly appreciated, and I look forward to hearing from the community.

**Comment by alexandercerutti** [comment]
Ho @akash-86, thank you for using passkit-generator!

I've never setup an add-to-wallet button, but I remember to have seen several details on Apple HIG.

Comes naturally that I don't know any libraries for such buttons.

For what I know, the idea behind is the one of generating the pass when the user clicks the button. Therefore you'll need to have and endpoint exposed to generate one on demand. You generally shouldn't generate the pass before the user request and then save it, because it might be useless.

Let me know if this helps or if you have any further questions.

**Reply by akash-86** [reply]
@alexandercerutti Thank you for your quick response. Now I have figured it out from your advice that I have to generate the Endpoint for the passes to make them user to be able to add it to their wallet.
- Can you please Share me any suggestion regarding how to change the font size or styles of the values which is present in pass such as secondary fields, auxiliary fields If possible ?
- What are the procedure for getting the notifications in my IOS device on any updates which is made in passes.

**Reply by alexandercerutti** [reply]
@akash-86

 - These kind of styles personalizations are not possible
 - You have to follow this document by Apple, which will give you an insight of what you should do. https://developer.apple.com/documentation/walletpasses/adding_a_web_service_to_update_passes
 
Furthermore, based on that document and the older one (available in documentation archives), I've created a typescript-ready toolkit that let you focus on the interaction with your system, without caring much about the protocol Apple establishes.

Right now only fastify implementation is available, but you can build an integration for a different webserver system (e.g. express) on top of the toolkit without any problem, as it exposes the endpoint, the structure etc.

https://github.com/alexandercerutti/passkit-webservice-toolkit
https://github.com/alexandercerutti/fastify-passkit-webservice

Give them a look, and let me know!

**Comment by akash-86** [comment]
So currently I am using latest node version for development purpose, but for the production requirements It will need to be shifted in node version 12 because the project is build upon it. So can passkit-generator will support version 12 ?

**Reply by alexandercerutti** [reply]
Node 12 was an LTS but its Active Support ended over 3 years ago (EOL). So, no, I won't test on Node 12.

I'm sure it will work on Node 14 (and it might work on Node 12 with some tweaks), as that was the latest LTS when I built passkit-generator. But yet, since Node 14-18 went EOL (18 reached it on Oct 2023), on the next big update I will make the jump to the most recent LTS.

And you should too for several reasons, security in the first place.

**Reply by akash-86** [reply]
I am pleased to inform you that I recently conducted tests with passkit-generator on Node 12 and, with a few adjustments, found it to be functioning seamlessly. Your guidance on Node 14 as the latest LTS is duly noted, and I'm inclined to align with your forward-looking approach.

On another note,  I'd like to inquire about Apple Wallet pass notifications. Does apple wallet passes supports any push notification thing so that a user can be notified for any updates or pass changes. Could you share any insights or best practices related to implementing notifications ?

Thank you for your time and expertise.

**Reply by alexandercerutti** [reply]
They support notifications, of course, but I don't have any best practice or whatever to follow, outside what I sent you in the previous days and outside of what can be found on Apple Developers Archive.

I never implemented a production-ready system, but as I sent you earlier in this conversation, I've developed two libraries to build one.

Hope this helps!
